### PR TITLE
Add support for hashmaps

### PR DESCRIPTION
**Pull request by kysre**
#598

### CODE REVIEW

**Comment by WisdomPill** [comment]
thanks for the effort @kysre 

can you merge master on your branch? just so tests might start to run and your PR will look cleaner?

**Comment by WisdomPill** [comment]
never mind I've fixed it

**Comment by kysre** [comment]
Do you suggest that I encode my keys, values like they are encoded in set() function?
```
nkey = self.make_key(key, version=version)
nvalue = self.encode(value)
```

**Comment by kysre** [comment]
I fixed the problem with encoding and decoding values. Now I am getting errors like below in tests:
```
________________________ TestDjangoRedisCache.test_hlen ________________________
  
  self = <tests.test_backend.TestDjangoRedisCache object at 0x7f4c20da6af0>
  cache = <django.utils.connection.ConnectionProxy object at 0x7f4c218c93a0>
  
      def test_hlen(self, cache: RedisCache):
  >       assert cache.hlen("foo_hash3") == 0
  
  tests/test_backend.py:819: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  django_redis/cache.py:29: in _decorator
      return method(self, *args, **kwargs)
  django_redis/cache.py:198: in hlen
      return self.client.hlen(*args, **kwargs)
  django_redis/client/default.py:860: in hlen
      client = self.get_client(write=False)
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  
  self = <django_redis.client.sharded.ShardClient object at 0x7f4c209f5a60>
  args = (), kwargs = {'write': False}
  
      def get_client(self, *args, **kwargs):
  >       raise NotImplementedError
  E       NotImplementedError
  django_redis/client/sharded.py:27: NotImplementedError
  ```
  Should I implement hashmap methods in `django_redis/client/sharded.py` too?

File: django_redis/client/default.py [hunk 1]
```diff
@@ -813,3 +813,79 @@ def touch(
...
+        self,
+        name: str,
+        client: Optional[Redis] = None,
+    ) -> List:
+        """
+        Return a list of keys in hash name.
+        """
+        if client is None:
+            client = self.get_client(write=False)
+        try:
```
Comment (WisdomPill) [inline django_redis/client/default.py:873]: why the try/catch?

File: django_redis/client/default.py [hunk 2]
```diff
@@ -813,3 +813,79 @@ def touch(
...
+        """
+        if client is None:
+            client = self.get_client(write=False)
+        return int(client.hlen(name))
+
+    def hkeys(
+        self,
+        name: str,
+        client: Optional[Redis] = None,
+    ) -> List:
```
Comment (WisdomPill) [inline django_redis/client/default.py:867]: list of what?

**Review by WisdomPill — CHANGES_REQUESTED** [review]
Overall looks good, just a couple of comments.

I think the problem with the tests you wrote is the `ShardedClient`, you can decide to skip those tests or implement them in the `ShardedClient`.

If you want your tests to work with shared client just change the `get_client` method with `make_key` and `get_server` based on the key

File: django_redis/client/default.py (hunk 1 above)
Comment (kysre) [reply django_redis/client/default.py:873]: I implemented hkeys() similar to [keys()](https://github.com/jazzband/django-redis/blob/9c8cf8f32b18bf57bd684ee305e9ba6198e4ef1e/django_redis/client/default.py#L726) function. If you think that the try/catch is redundant, I can remove it.

File: django_redis/client/default.py (hunk 2 above)
Comment (kysre) [reply django_redis/client/default.py]: Fixed it.

**Comment by kysre** [comment]
I decided to skip `ShardedClient` tests for now. If you think it's better to implement them in `ShradedClient`, tell me and I'll do it.

File: django_redis/client/default.py (hunk 1 above)
Comment (WisdomPill) [reply django_redis/client/default.py:873]: it's okay then, it was implemented way before I took this project, maybe we should discuss it in a separate time

**Comment by WisdomPill** [comment]
it's okay, I would merge it and then add TODOs in the issue so that other people can take it a step further and add support other commands methods too

**Pull request merged**